Optionally skip initial adaptivity projections

roystgnr opened the pull request:
This is a successful workaround for #25007 for me; the deeper problems uncovered are going to require libMesh fixes.

I can't enable this behavior by default without exodiffs of marker variables in a ton of tests; I'm leaning toward valuing "backwards compatibility" (at least for users' test suites; not sure how many people care to retain those projected markers) over "efficiency without setting a flag" for now.

We now have a better short-term workaround for the specific case prompting #25007, but long-term I think this may still be a valuable option to have.  Even a performance-bug-free initial projection is never going to be as cheap as no-projection-at-all, and for most users no-(mesh-to-mesh)-projection-at-all is going to be a fine option when variables are just going to be immediately set again by IC inputs.

roystgnr committed:
Optionally skip initial adaptivity projections

This is a successful workaround for #25007 for me; the deeper problems
uncovered are going to require libMesh fixes.

I can't enable this behavior by default without exodiffs of marker
variables in a ton of tests; I'm leaning toward valuing "backwards
compatibility" (at least for users' test suites; not sure how many
people care to retain those projected markers) over "efficiency without
setting a flag" for now.

lindsayad reviewed (changes requested):
Will need a test to get coverage since the default is to project

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1671280/) on 1f790f6 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25067/site)

This comment will be updated on new commits.

GiudGiud commented on framework/src/actions/SetAdaptivityOptionsAction.C:
```diff
@@ -33,6 +33,9 @@ SetAdaptivityOptionsAction::validParams()
       "The name of the Marker to use to adapt the mesh during initial refinement.");
+  params.addParam<bool>("project_initial_marker",
+                        true,
+                        "Whether to project initial Marker values onto the refined mesh.");
```
can we have more context here?

Is it needed to project to run 2 rounds of adaptivity? so for a single level we dont need to?

roystgnr replied:
As far as the framework is concerned, it's never needed.  libMesh doesn't even write a marker to a variable, so it certainly doesn't care; we get our refinement flag from `Elem::refinement_flag`, not a constant `MONOMIAL` coefficient.  MOOSE writes markers to variables, but once the Adaptivity system itself reads the marker there's nothing in a MOOSE run by default that cares afterward.  The efficient thing to do would have been to only write the marker to a variable in the first place in those situations where it's explicitly requested or where another object uses it later.  And I'd bet that even in those situations, "later" is probably just when another Marker uses the output of the first, and we still only need that output before a refinement, not after.

There are two catches, though.  Number 1 is that users can do whatever they want; it's entirely possible for someone to throw on some postprocessor that reads a marker variable, for example, and defaulting to `false` here would break that.  Number 2 is that everything we write to a variable gets written out to restart and visualization output, and defaulting to `false` here basically breaks every `exodiff` test with initial adaptivity.

I'm not actually sure where to put all this info.  It's too much for an `addParam` blurb, but https://github.com/idaholab/moose/blob/next/framework/doc/content/source/base/Adaptivity.md is basically an empty shim at the moment.  Maybe now's the time to fix that, or maybe there's some other doc I'm not finding?

GiudGiud replied:
https://mooseframework.inl.gov/moose/syntax/Adaptivity/

GiudGiud replied:
All those exodiffs can be avoided by setting the parameter in the test btw. It s not a concern to me

lindsayad replied:
I do think it would be nice for the default to be `true` ... it would be interesting to see how many apps in CI that breaks.

This makes me think of https://github.com/idaholab/moose/pull/12389 where CI essentially beat me into submission, with the result being that a seemingly inferior default stays the default

roystgnr committed:
Try out a default project_initial_marker=false

We'll see whether manually specifying `true` keeps existing MOOSE CI
happy, as well as whether we have further problems in downstream CI.

recuero commented:
> at least for users' test suites; not sure how many people care to retain those projected markers

The contact test was only there to ensure initial adaptivity worked with some neighbor projection settings, so they do not rely on any kind of reference solution. I'd also be fine without the backwards compatibility since, as far as I read, the new method is more accurate. I'd imagine diffs aren't huge.

roystgnr committed:
Some more project_initial_marker overrides

I was grepping for `initial_marker`, but we do initial adaptivity on
`marker` too if the more specific name isn't set.

roystgnr committed:
Typo fix

Didn't need to update the test file here, but I was quite confused when
the expected `--re` didn't work to verify that.
